Fix compilation on Ubuntu Linux

0mp opened the pull request:
This PR solves a couple of problems I stumbled upon as I was building OpenBSM on Ubuntu 17.04:

1. `getgrouplist()` is not defined in `unistd.h` on Linux as it is on FreeBSD. It can be found in `grp.h` though.
2.  `pthread_mutex_isowned_np` is not available on Linux. It is never used, however, so it is possible to ignore it when compiling on Linux.
3. Yacc and flex (or any other compatible replacements) are required for the compilation to succeed. Hence, I added a small note to the INSTALL file about it. 
 
   At the moment the note suggests to install `bison`. Should I change it to `byacc`?

cemeyer commented:
Looks ok to me.

cbrueffer commented on INSTALL:
```diff
@@ -43,3 +43,9 @@ directory the correct libbsm is used:
     LD_LIBRARY_PATH=/home/rwatson/openbsm/libbsm ; export LD_LIBRARY_PATH
+
+In order to compile auditdistd it might be necessary to get yacc or a
+compatible program. Also, it is necessary to have a lexcial analyser generator
```
lexcial -> lexical

cbrueffer commented on INSTALL:
```diff
@@ -43,3 +43,9 @@ directory the correct libbsm is used:
+compatible program. Also, it is necessary to have a lexcial analyser generator
+such as flex. The following appears to work on Ubuntu:
+
+    sudo apt-get install bison flex
```
This should probably match our Travis config; currently we use byacc instead of bison there. Any preference for one or the other?

0mp replied:
@cbrueffer License maybe? Byacc is public domain software whereas Bison is GPL.

0mp replied:
I changed it to byacc.

0mp commented:
It's worth noting that similar messages show up during compilation on macOS. The difference is that they are treated as warnings (in the `pthread_mutex_isowned_np` case). Apparently, the GCC configuration on Ubuntu Linux is less forgiving.

Is conditional compilation really a solution here?

cbrueffer commented:
Thanks for the updates! Not sure about the conditional compilation, maybe @rwatson wants to chime in?

rwatson commented:
It looks like the use of ```pthread_mutex_isowned_np``` is for an unused debugging function in ```auditdistd```'s ```synch.h```, which contains a number of other utility functions to allow code written against the FreeBSD kernel's ```mutex.h``` to be implemented using pthreads. Presumably this is so that the code could be used in the kernel or userspace, although in practice it is used only in userspace. I think it would be most reasonable to detect whether ```pthread_mutex_isowned_np``` is present, and conditionally compile its calling function based on that, although this is slightly more work than simply deleting the calling function. I suggest pursuing this route because, in maintaining the code in the future, we may want further locking assertions, for which this is useful.

(There is a more broad question about whether we should rewrite the ```auditdistd`` code to use pthreads in preference to the kernel APIs. I think if I were writing this code from scratch, and given what we now know about where it is used, I would prefer pthreads there. But I'm not sure I see urgency in changing the current code.)

0mp commented:
@rwatson 

> I think it would be most reasonable to detect whether `pthread_mutex_isowned_np` is present, and conditionally compile its calling function based on that, although this is slightly more work than simply deleting the calling function.

Isn't it already done in this patch? `mtx_owned` is the function calling `pthread_mutex_isowned_np` and it is conditionally compiled now (using `#if defined(HAVE_PTHREAD_MUTEX_ISOWNED_NP)`).

Would you prefer to see an implementation of `pthread_mutex_isowned_np` for Linux instead?

csjayp commented on bin/auditdistd/sandbox.c:
```diff
@@ -45,6 +45,9 @@
 #include <stdlib.h>
 #include <strings.h>
 #include <unistd.h>
+#if defined(__linux__)
```
I would add a comment here that we are conditionally including this header in Linux because getgrouplist() and setgroups() isn't defined in unistd.h (since grp.h exists in both Linux and BSD)

csjayp replied:
In general I think this PR looks good but we need to fix CI. It looks like we might need to specify `-I` and `-L`  for `/usr/local/opt/openssl ...` (something specific to homebrew). It looks like the build doesn't know where to find openssl

0mp replied:
I'll probably have to rebase this branch in order for the CI to work. Hang on!

0mp commented:
I've rebased the branch to fix the CI.

I've also squashed 3 commits referencing the following issues raised in the comments:

 1. https://github.com/openbsm/openbsm/pull/12#discussion_r131135564
 2. https://github.com/openbsm/openbsm/pull/12#discussion_r131136049
 3. https://github.com/openbsm/openbsm/pull/12#discussion_r165559287

@csjayp the CI seems to work now and I added the comment as you requested.
